File: test_libhid.c
#include <hid.h>
#include <stdio.h>
#include <string.h>
#include <unistd.h> /* for getopt() */

bool match_serial_number(struct usb_dev_handle* usbdev, void* custom, unsigned int len)
{
  bool ret;
  char* buffer = (char*)malloc(len);
  usb_get_string_simple(usbdev, usb_device(usbdev)->descriptor.iSerialNumber,
      buffer, len);
  ret = strncmp(buffer, (char*)custom, len) == 0;
  free(buffer);
  return ret;
}

void set_debug(){
  /* see include/debug.h for possible values */
  hid_set_debug(HID_DEBUG_WARNINGS);
  hid_set_debug_stream(stderr);
  /* passed directly to libusb */
  hid_set_usb_debug(0);
}

int do_hid_init(){
  hid_return ret;
  fprintf(stderr, "Before hid_init\n");
  ret = hid_init();
  if (ret != HID_RET_SUCCESS) {
    fprintf(stderr, "hid_init failed with return code %d\n", ret);
    return 1;
  }
  fprintf(stderr, "After hid_init\n");
  return 0;
}

int force_open_interface(HIDInterface * hid, HIDInterfaceMatcher matcher, int iface_num){
  hid_return ret;
  ret = hid_force_open(hid, iface_num, &matcher, 5);
  fprintf(stderr, "After hid_force_open %d\n", iface_num);
  if (ret != HID_RET_SUCCESS) {
    fprintf(stderr, "hid_force_open failed with return code %d\n", ret);
    return 1;
  }
  return 0; 
}



int main(int argc, char *argv[])
{
  HIDInterface* hid_interface_0;
  HIDInterface* hid_interface_1;
  hid_return ret;

  unsigned short vendor_id  = 0x1130;
  unsigned short product_id = 0x0202;
  char *vendor, *product;

  int flag;

  HIDInterfaceMatcher matcher = { vendor_id, product_id, NULL, NULL, 0 };

  set_debug();  
  if(do_hid_init() == 1)
      return 1;

  hid_interface_0 = hid_new_HIDInterface();
  if (hid_interface_0 == 0) {
    fprintf(stderr, "hid_new_HIDInterface() failed, out of memory?\n");
    return 1;
  }
  fprintf(stderr, "After created new HIDInterface hid_interface_0 \n");

  hid_interface_1 = hid_new_HIDInterface();
  if (hid_interface_1 == 0) {
    fprintf(stderr, "hid_new_HIDInterface() failed, out of memory?\n");
    return 1;
  }
  fprintf(stderr, "After created new HIDInterface hid_interface_1\n");

   if(force_open_interface(hid_interface_0, matcher, 0) == 1)
        return 1;
   if(force_open_interface(hid_interface_1, matcher, 1) == 1)
        return 1;
 
   char write_cbw_packet[8] = {0x55, 0x53, 0x42, 0x43, 0, 16, 2,0};
   unsigned char const PATHLEN = 1;
   int const PATH_IN[1]  =  { 4194304 };
   int const PATH_OUT[3] = { 0xffa00001, 0xffa00002, 0xffa10004 };
 
   fprintf(stderr, "Before hid_set_output_report\n");
   ret = hid_set_output_report(hid_interface_1, PATH_IN, PATHLEN, "", 8);
   ret = hid_interrupt_write(hid_interface_0, 0x81, write_cbw_packet, 8, 1000); 
   ret = hid_set_output_report(hid_interface_0, PATH_IN, PATHLEN, "", 16);

   printf("\n%d\n\n", ret); // Prints out '0'



   char green_light[16] = {0x5a,0x5a,0x5a,0x5a,
                           0x5a,0x5a,0x5a,0x5a,
                           0x5a,0x5a,0x5a,0x5a,
                           0x5a,0x5a,0x5a,0x5a};


   fprintf(stderr, "Before hid_interrupt_write again:\n");
   ret = hid_interrupt_write(hid_interface_0, 0x81, green_light, 16, 1000); 
   printf("\n%d\n", ret); // Prints out '21' 

  if (ret != HID_RET_SUCCESS) {
    fprintf(stderr, "hid_cleanup failed with return code %d\n", ret);
    return 1;
  }

  ret = hid_close(hid_interface_0);
  ret = hid_close(hid_interface_1);

  hid_delete_HIDInterface(&hid_interface_0);
  hid_delete_HIDInterface(&hid_interface_1);

  ret = hid_cleanup();
  
  return 0;
}

/* COPYRIGHT --
 *
 * This file is part of libhid, a user-space HID access library.
 * libhid is (c) 2003-2005
 *   Martin F. Krafft <libhid@pobox.madduck.net>
 *   Charles Lepple <clepple@ghz.cc>
 *   Arnaud Quette <arnaud.quette@free.fr> && <arnaud.quette@mgeups.com>
 * and distributed under the terms of the GNU General Public License.
 * See the file ./COPYING in the source distribution for more information.
 *
 * THIS PACKAGE IS PROVIDED "AS IS" AND WITHOUT ANY EXPRESS OR IMPLIED
 * WARRANTIES, INCLUDING, WITHOUT LIMITATION, THE IMPLIED WARRANTIES
 * OF MERCHANTIBILITY AND FITNESS FOR A PARTICULAR PURPOSE.
 */

File: test_libhid.py
#!/usr/bin/python
#
# This is just an example. Please see the test_libhid.c file for more
# comments.
#

CBW = "/0x55/0x53/0x42/0x43/0/16/2/0/"

import sys

# allow it to run right out of the build dir
import os
libsdir = os.getcwd() + '/.libs'
if os.path.isdir(libsdir) and os.path.isfile(libsdir + '/_hid.so'):
  sys.path.insert(0, libsdir)

from hid import *

def open_interface(interface, matcher):
  hid = hid_new_HIDInterface()
  ret = hid_force_open(hid, interface, matcher, 3)
  if ret != HID_RET_SUCCESS:
    sys.stderr.write("hid_force_open %s failed with return code %d.\n" % (interface, ret))
  return hid

def close_interface(hid):
  ret = hid_close(hid)
  if ret != HID_RET_SUCCESS:
    sys.stderr.write("hid_close failed with return code %d.\n" % ret)

def main():
  hid_set_debug(HID_DEBUG_WARNINGS)
  hid_set_debug_stream(sys.stderr)
  hid_set_usb_debug(0)

  ret = hid_init()

  matcher = HIDInterfaceMatcher()
  matcher.vendor_id = 0x1130
  matcher.product_id = 0x0202

  hid_interface_1 = open_interface(1, matcher)
  hid_interface_0 = open_interface(0, matcher)
  sys.exit() 
  # Set_Report with report type output, report ID 0, interface 1, and length 8

  ret = hid_set_output_report(hid_interface_1, (), "8") 

  # OUT transaction to endpoint 0 with CBW indicates write and length.
  ret = hid_interrupt_write(hid_interface_0, 0x81, CBW, 1000);

  # Set_Report with report type output, report ID 0, interface 0, length 16.

  ret = hid_set_output_report(hid_interface_0, (), "16") 

  # OUT transaction to endpoint 0 with data and length 8.

  ret = hid_interrupt_write(hid_interface_0, 0x81, "/0xff/0xfe/0xfd/0xfc/0xfb/0xfa/0xf9/0xf9", 1000);

  # OUT transaction to endpoint 0 with data and length 8.

  ret = hid_interrupt_write(hid_interface_0, 0x81, "/0xf7/0xf6/0xf5/0xf4/0xf3/0xf2/0xf1/0xf0", 1000);

  close_interface(hid_interface_1)
  
  hid_cleanup()

if __name__ == '__main__':
  main()

# COPYRIGHT --
#
# This file is part of libhid, a user-space HID access library.
# libhid is (c) 2003-2005
#   Martin F. Krafft <libhid@pobox.madduck.net>
#   Charles Lepple <clepple@ghz.cc>
#   Arnaud Quette <arnaud.quette@free.fr> && <arnaud.quette@mgeups.com>
# and distributed under the terms of the GNU General Public License.
# See the file ./COPYING in the source distribution for more information.
#
# THIS PACKAGE IS PROVIDED "AS IS" AND WITHOUT ANY EXPRESS OR IMPLIED
# WARRANTIES, INCLUDING, WITHOUT LIMITATION, THE IMPLIED WARRANTIES
# OF MERCHANTIBILITY AND FITNESS FOR A PARTICULAR PURPOSE.
